src/dmd/semantic3.d

```diff
@@ -241,6 +241,20 @@ private extern(C++) final class Semantic3Visitor : Visitor
         {
             if (funcdecl.storage_class & STC.inference)
             {
+                /* https://issues.dlang.org/show_bug.cgi?id=18672
+                   If the parent struct has a destructor that is
+                   `@safe`, then the `opAssign` function is `@trusted`
+                 */
+                auto sd = funcdecl.parent.isStructDeclaration();
+                if (auto dtor = sd.dtor)
+                {
+                    if (dtor.isSafe())
```

Comment:
Please correct me if I'm wrong, but I don't think it should check if the destructor is `@safe`, but that the context in which the struct is declared is `@safe`.

Reply:
To elaborate, marking something as `@trusted` involves manual auditing.  How do we know that `opAssign` is memory-safe in this context.  And what if `opAssign` is written by the user?  Will that be automatically annotated as `@trusted`?

Reply:
No, this test is solely for the generated opAssign. We know how the generated opAssign looks, because, well, the compiler generates it [1]. The body of the generated function looks like this:

```d
S __swap = void;
 __swap = this;   // bit copy
 this = s;        // bit copy
 __swap.dtor();
```

As you can see, the `S __swap = void` is unsafe, but it can be trusted. So if the destructor is @safe, we can trust the whole opAssign.

[1] https://github.com/dlang/dmd/blob/master/src/dmd/clone.d#L210

Reply:
>Please correct me if I'm wrong, but I don't think it should check if the destructor is @safe, but that the context in which the struct is declared is @safe

I see. I, wrongfully, thought that you can declare a `@system` function inside a `@safe` one. For example, on today's compiler, I can do this:

```d
void delegate() fp; 

void fun2() @safe
{

    void fun()
    {
        int a;
        int* b = &a;                  // totally unsafe                                                                                        
        import std.stdio : writeln;
        writeln("called system function inside safe context");
    }

    fp = &fun;
}

void main()
{
    fun2();
    fp();
}
```

However, the spec is very clear on this one: https://dlang.org/spec/function.html#safe-functions (point number 4). So it seems that the bug is not about `opAssign`, but a more general one: if the scope is @safe, then function declarations inside that scope should be @safe.

Reply:
After reading the spec more carefully, I could unveil the true nature of this bug: dmd fails to propagate safety to nested functions. It seems that attributes are inferred solely for templates and function literals, so the generated opAssign body should not be analyzed in order to establish if it is @safe or not. Therefore, the current fix is wrong and I will try to fix the bigger problem, which is propagating safety to nested functions. In the case of other attributes, the spec is unclear whether those should be propagated as well or not; I would assume that all attributes should be treated the same.

Reply:
> It seems that attributes are inferred solely for templates and function literals

Yeah, that’s excepted, at least for user defined functions.